src/mod_domain_isolation.erl

```diff
@@ -0,0 +1,73 @@
+-module(mod_domain_isolation).
+-behaviour(gen_mod).
+
+-include_lib("exml/include/exml.hrl").
+-include_lib("jid/include/jid.hrl").
+-include("mongoose_config_spec.hrl").
+
+%% gen_mod handlers
+-export([start/2, stop/1]).
+-export([config_spec/0]).
+
+%% hooks
+-export([filter_local_packet/1]).
+
+-type fpacket() :: {From :: jid:jid(),
+                    To :: jid:jid(),
+                    Acc :: mongoose_acc:t(),
+                    Packet :: exml:element()}.
+
+-spec config_spec() -> mongoose_config_spec:config_section().
+config_spec() ->
+    #section{
+       items = #{<<"extra_domains">> => extra_domains()}
+      }.
+
+extra_domains() ->
+    #list{items = #option{type = string, validate = domain_template}}.
+
+start(Host, Opts) ->
+    ExtraDomains = proplists:get_value(extra_domains, Opts, []),
+    SubHosts = [gen_mod:make_subhost(SubHost, Host) || SubHost <- ExtraDomains],
+    AllHosts = [Host|SubHosts],
+    [ejabberd_hooks:add(filter_local_packet, H, ?MODULE, filter_local_packet, 10)
+     || H <- AllHosts],
+    ok.
+
+stop(Host) ->
+    ExtraDomains = gen_mod:get_module_opt(Host, ?MODULE, extra_domains, []),
+    SubHosts = [gen_mod:make_subhost(SubHost, Host) || SubHost <- ExtraDomains],
+    AllHosts = [Host|SubHosts],
+    [ejabberd_hooks:delete(filter_local_packet, H, ?MODULE, filter_local_packet, 10)
+     || H <- AllHosts],
+    ok.
+
+-spec filter_local_packet(Value :: fpacket() | drop) -> fpacket() | drop.
+filter_local_packet(drop) ->
+    drop;
+filter_local_packet({From = #jid{lserver = FromServer},
+                     To = #jid{lserver = ToServer}, Acc, Packet}) ->
```

Comment:
Won't it happen that if we can be sure both `FromServer` and `ToServer` are the same, including subdomains, then this is allowed? In that case we could do
```suggestion
filter_local_packet({From = #jid{lserver = SameServer},
                     To = #jid{lserver = SameServer}, Acc, Packet}) ->
    {From, To, Acc, Packet};
filter_local_packet({From = #jid{lserver = FromServer},
                     To = #jid{lserver = ToServer}, Acc, Packet}) ->
```
as a shortcut path to skip extracting domains?